content/articles/protobuf-instead-of-json-in-frontend.md

```diff
@@ -0,0 +1,107 @@
+
+- smaller in size
+- typed
+- offer a common interface for all your services, you can just update your .proto or .thrift and share those with the services using it, as long as there is a client for that language
+
+[Thrift](https://thrift.apache.org/) does more as it also allows you to define methods as well as structs.
+
+## Example app
+As an example we are going to make a small address book app. Coincidentally that's the example that the Python Protobuf doc is using as well but that wasn't on purpose I swear! The data is generated randomly on startup of the flask app.  
+The repo is on our [github](https://github.com/WeAreWizards/protojson) if you want to try it yourself.  
+As mentioned on the README, the Python client for Protobuf does not support Python3 (yet) so you will have to make a Python2 virtualenv if you decide to install the server.  
+The server will deliver the content in JSON or Protobuf based on the `Accept` header.
+
+### Server-side
+Let's have a quick look at the server, it might not be new if you have already used Protobuf before but it is helpful for the rest (like me, I only used Thrift before).  
+First we need to define our Protobuf schema, which is done in the [addressbook.proto](https://github.com/WeAreWizards/protojson/blob/master/addressbook.proto) file. Here is one of the model as example:
+
+```protobuf
+message Contact {
+  required string first_name = 1;
+  required string last_name = 2;
+  optional Address address = 3;
+  repeated Phone phone_numbers = 4;
+};
+```
+In this model, `first_name` and `last_name` are required (controversial, I'll come back to that later), the address model is optional and we can have 0, 1 or n `phone_numbers`.  
+Now, Python doesn't know about .proto files so we need to use the [python client](https://pypi.python.org/pypi/protobuf) to generate code that we can use. This is the [addressbook_pb2.py](https://github.com/WeAreWizards/protojson/blob/master/addressbook_pb2.py) file. You can then import that in your Python code and access the models defined in your .proto file.  
+Have a look on the [data.py](https://github.com/WeAreWizards/protojson/blob/master/data.py) client to see an example. You can use those models as you would with a class and using kwargs for every attributes.  
+You also need to serialize them before sending, using the `SerializeToString()` method in our app.  
+
+### Client-side
+The client is written in Angular purely as an example and because ng-repeat was exactly what I wanted here. The code is a quick prototype so don't expect it to be idiomatic.  
+It uses [ProtoBuf.js](https://github.com/dcodeIO/ProtoBuf.js) as a Protobuf client.  
+The first thing we do is a basic GET to get all our contacts which is trivial since you can just dump the response data into our scope variable. On the other hand, we need to handle Protobuf deserialization since looking at our network tab we receive binary data looking like:
```

Comment:
Is the reason you can just dump your response that angular has a default json decoding implementation?